xdem/profiler.py

```diff
@@ -0,0 +1,385 @@
+
+        # Plot memory usage line
+        fig = go.Figure()
+        fig.add_trace(go.Scatter(x=times, y=values, mode="lines+markers", name=f"{data_name} usage"))
+
+        # Base Y position for markers
+        base_y = max(values)
+        offset_step = (max(values) - min(values)) / 50  # how much higher each subsequent label goes
+        current_offset = -offset_step * 2
+
+        for _, row in list(subcalls.iterrows())[::-1]:
+            sub_t = row["call_time"] - call_start_time
+            sub_name = row["name"]
+
+            # start function
+            y_position = base_y + current_offset
+            fig.add_trace(
+                go.Scatter(
+                    x=[sub_t],
+                    y=[y_position],
+                    mode="markers+text",
+                    marker={
+                        "color": "black",
+                        "size": 8,
+                    },
+                    text=[sub_name],
+                    textposition="middle right",  # text right next to marker at same height
+                    showlegend=False,
+                )
+            )
+
+            fig.add_shape(
+                type="line",
+                x0=sub_t,
+                x1=sub_t,
+                y0=min(values),
+                y1=y_position,
+                line={
+                    "color": "black",
+                    "width": 1,
+                    "dash": "dot",
+                },
+            )
+
+            # Increment offset for end function
+            current_offset += offset_step
+            y_position = base_y + current_offset
+
+            fig.add_trace(
+                go.Scatter(
+                    x=[sub_t + row["time"]],
+                    y=[y_position],
+                    mode="markers+text",
+                    marker={
+                        "color": "black",
+                        "size": 8,
+                    },
+                    text=["end " + sub_name],
+                    textposition="middle right",  # text right next to marker at same height
+                    showlegend=False,
+                )
+            )
+
+            fig.add_shape(
+                type="line",
+                x0=sub_t + row["time"],
+                x1=sub_t + row["time"],
+                y0=min(values),
+                y1=y_position,
+                line={
+                    "color": "black",
+                    "width": 1,
+                    "dash": "dot",
+                },
+            )
+
+            # Increment offset for next marker
+            current_offset += offset_step
+
+        fig.update_layout(
+            title="{} usage during {} call".format(data_name, parent_row["name"]),
+            xaxis_title="Time (s)",
+            yaxis_title="Memory (MB)",
+            showlegend=True,
+        )
+
+        return fig
+
+
+def profile(name: str, interval: int | float = 0.05, memprof: bool = False):  # type: ignore
+    """
+    xDEM profiling decorator
```

Comment:
This will be one of the public functions (in the API) for users, so we should describe things as thoroughly as we can directly here (even if redundant to the documentation).
We can simply copy the description + usage example you already wrote in "**Modifying the profiled functions**" section of the doc :wink: